src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import data.fintype.basic
+import algebra.big_operators.basic
+import data.equiv.fin
+
+/-!
+# The Hales-Jewett theorem
+
+We prove the Hales-Jewett theorem and deduce Van der Waerden's theorem as a corollary.
+
+We consistently work with finite types instead of natural numbers. That is, we write `m, n, k` for
+(finite) types where one might traditionally take `m, n, k` to be natural numbers. This allows us to
+work directly with `m`, `option m`, `(n → m) → k`, and `n ⊕ n'` instead of `fin m`, `fin (m+1)`,
+`fin (k^(m^n))`, and `fin (n + n')`.
+
+We take '`X` is monochromatic' to mean 'there exists a color `c` such that every point of `X` has
+the color `c`' as opposed to 'any `x y : X` have the same color' since the former is generally
+easier to work with. This means that our formulation of Hales-Jewett would be false when `m` and `k`
+are both empty (then there can be a coloring even though there is no color). Thus we require `m` to
+be nonempty in the statement of `exists_mono_in_high_dimension`.
+
+We do not prove the finitary version of Van der Waerden's theorem (which is equivalent to the
+infinite one by compactness), although it follows from the same proof.
+
+In principle our proof should give computable upper bounds on the numbers `HJ(m, k)`, but this would
+require rewriting the proof to avoid `Prop`, and the upper bounds are enormous anyway.
+
+## Main results
+
+- `combinatorics.line.exists_mono_in_high_dimension m k`: the Hales-Jewett theorem.
+- `combinatorics.exists_mono_homothetic_copy M S f k C`: a generalization of Van der Waerden's
+theorem to an arbitrary commutative monoid. The proof is a simple application of Hales-Jewett.
+
+## Tags
+
+combinatorial line, Ramsey theory, arithmetic progession
+
+### References
+
+* https://en.wikipedia.org/wiki/Hales%E2%80%93Jewett_theorem
+
+Note that this file does not use the same notation as Wikipedia.
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+instance (n : pnat) : nonempty (fin n) := ⟨⟨0, n.pos⟩⟩
+
+/-- An induction principle for types analogous to `pnat.rec_on`. To prove a statement about
+all nonempty finite types, it suffices to consider `unit` and `option α` where we know the
+statement for `α`, provided that our statement is preserved by `≃`.-/
+lemma fintype_induction {P : Type → Prop} (of_equiv : ∀ {α β}, α ≃ β → P α → P β)
```

Comment:
Is there any need to restrict to `Type` rather than `Type*` throughout? As far as I can tell the only times you actually need it to be `Type` are because of `fin` and `pnat`, but you could use `ulift` to generalise the main HJ theorem

Reply:
I think it's nice to use `Type` in the proof of HJ because this way `empty`, `unit`, `κ`, and `(ι → α) → κ` all lie in the correct universe.  One could deduce a universe-polymorphic version of HJ from the `Type`-version since any fintype bijects with `fin n`.

Reply:
You could adjust `empty` and `unit` to `pempty` and `punit` respectively, then I think the four types you mention still lie in the correct universe?

Reply:
The problem is if `α` has a higher universe level than `κ`. Then `(ι → α) → κ` is "too big". It would make sense to have a version of HJ with arbitrary universe levels, but I think it's easiest to prove this via the current restricted version (using `α ≃ fin (fintype.card α)`).

Reply:
I added fully polymorphic versions of HJ and VdW, it was easier than expected.